### ISSUE TITLE
CheriABI top(1) crashes with tag violation

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/cheribsd#203]
With a CheriABI ```top(1)```, I see the follow warning message about termcap (possibly an unrelated problem), and then this crash:
```
root@qemu-cheri:~ # top
top: no termcap entry for a `vt100' terminal
CHERI cause: ExcCode: 0x02 RegNum: $c01 (tag violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c02: v:1 s:0 p:7fff807d b:000000000049f040 l:0000000000000040 o:0 t:0
$c03: v:1 s:0 p:7fff807d b:000000000049f080 l:0000000000000040 o:0 t:0
$c04: v:1 s:0 p:7fff807d b:000000000049f0c0 l:0000000000000040 o:0 t:0
$c05: v:1 s:0 p:7fff807d b:0000007fffffd298 l:0000000000000008 o:0 t:0
$c06: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c07: v:1 s:0 p:7fff807d b:0000007fffffcd38 l:0000000000000001 o:0 t:0
$c08: v:1 s:0 p:7fff807d b:0000000120246734 l:0000000000000004 o:0 t:0
$c09: v:1 s:0 p:7fff807d b:000000012024673c l:0000000000000004 o:0 t:0
$c10: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:120252660 t:0
$c11: v:1 s:0 p:7fff807d b:0000007fff7ff000 l:00000000007ff3c0 o:0 t:0
$c12: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:12001eee0 t:0
$c13: v:1 s:0 p:00008055 b:0000007fffffd340 l:00000000000000a0 o:0 t:0
$c14: v:1 s:0 p:7fff807d b:000000012026e580 l:0000000000000008 o:0 t:0
$c15: v:1 s:0 p:7fff807d b:000000012026e600 l:0000000000000008 o:0 t:0
$c16: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:12026e500 t:0
$c17: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:12000f58c t:0
$c18: v:1 s:0 p:7fff807d b:0000007fffffe0e0 l:0000000000000100 o:0 t:0
$c19: v:1 s:0 p:7fff807d b:0000000120258ca8 l:0000000000000008 o:0 t:0
$c20: v:1 s:0 p:7fff807d b:0000000120258d00 l:0000000000000004 o:0 t:0
$c21: v:1 s:0 p:7fff807d b:0000000120258d60 l:0000000000000020 o:0 t:0
$c22: v:1 s:0 p:7fff807d b:0000000120258ce0 l:0000000000000020 o:0 t:0
$c23: v:1 s:0 p:7fff807d b:0000000120258d20 l:0000000000000020 o:0 t:0
$c24: v:1 s:0 p:7fff807d b:0000000120258d40 l:0000000000000020 o:0 t:0
$c26: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:12001ef1c t:0
Mar 21 15:12:24 qemu-cheri kernel: USER_CHERI_EXCEPTION: pid 658 tid 100061 (top), uid 0: CP2 fault (type 0x32)
Mar 21 15:12:24 qemu-cheri kernel: Trapframe Register Dump:
Mar 21 15:12:24 qemu-cheri kernel: zero: 0      at: 0   v0: 0   v1: 0x8
Mar 21 15:12:24 qemu-cheri kernel: a0: 0x120258d20      a1: 0x120258d20 a2: 0x1a3: 0
Mar 21 15:12:24 qemu-cheri kernel: a4: 0        a5: 0   a6: 0   a6: 0x100000000
Mar 21 15:12:24 qemu-cheri kernel: t0: 0        t1: 0   t2: 0   t3: 0x120270000
Mar 21 15:12:24 qemu-cheri kernel: t8: 0xc      t9: 0x1200c4db8 s0: 0   s1: 0x1
Mar 21 15:12:24 qemu-cheri kernel: s2: 0xa0     s3: 0   s4: 0x12025a740 s5: 0xffd7
Mar 21 15:12:24 qemu-cheri kernel: s6: 0x12025a740      s7: 0x120260000 k0: 0  k1: 0
Mar 21 15:12:24 qemu-cheri kernel: gp: 0x12025a740      sp: 0x7fe320    s8: 0x7fe320    ra: 0x5
Mar 21 15:12:24 qemu-cheri kernel: sr: 0x408084b3       mullo: 0x3680000000000000       mulhi: 0        badvaddr: 0x12001ef1c
Mar 21 15:12:25 qemu-cheri kernel: cause: 0x48  pc: 0x12001ef1c
Signal 34 (core dumped)
```

The value in $epcc appears to correspond to this code in the ```percentages()``` function:
```
000000012001eee0 <percentages>:
int *out;
register long *new;
register long *old;
long *diffs;

{
   12001eee0:   67bdffe0        daddiu  sp,sp,-32
   12001eee4:   ebcbe81b        csd     s8,sp,24(c11)
   12001eee8:   03a0f025        move    s8,sp
    /* initialization */
    total_change = 0;
    dp = diffs;

    /* calculate changes for each state and the overall change */
    for (i = 0; i < cnt; i++)
   12001eeec:   1880002f        blez    a0,12001efac <percentages+0xcc>
   12001eef0:   00000000        nop
   12001eef4:   64020000        daddiu  v0,zero,0
   12001eef8:   64030008        daddiu  v1,zero,8
   12001eefc:   00802825        move    a1,a0
   12001ef00:   49a13000        cmove   c1,c6
    {
        if ((change = *new - *old) < 0)
   12001ef04:   c8240003        cld     at,zero,0(c4)
   12001ef08:   c8c50003        cld     a2,zero,0(c5)
   12001ef0c:   0026082f        dsubu   at,at,a2
   12001ef10:   28260000        slti    a2,at,0
        {
            /* this only happens when the counter wraps */
            change = (int)
   12001ef14:   00013800        sll     a3,at,0x0
    dp = diffs;

    /* calculate changes for each state and the overall change */
    for (i = 0; i < cnt; i++)
    {
        if ((change = *new - *old) < 0)
   12001ef18:   00e6080b        movn    at,a3,a2
        {
            /* this only happens when the counter wraps */
            change = (int)
                ((unsigned long)*new-(unsigned long)*old);
        }
        total_change += (*dp++ = change);
   12001ef1c:   e8210003        csd     at,zero,0(c1)
```

$c1 is NULL (and originates in $c6, which is also NULL), hence being untagged. It's unclear to me whether $c6, which I take to be the ```diffs``` argument, should in fact be NULL, possibly pointing at a bug in the caller. Will investigate further.

**Comment by rwatson** [comment]
This appears to be due to an incomplete prototype for the function ```percentages()``` in top's FreeBSD-specific ```machine.c```. As a result, a surprisingly NULL capability is passed as an ```diffs``` argument to the function. Now testing a patch for this.

**Comment by rwatson** [comment]
@davidchisnall: As discussed previously, as we're likely unable to generate code that will ever work with incomplete function prototypes when using pure-capability code, it would be great if compilation of such code always generated a compilation failure.

**Comment by rwatson** [comment]
With the fix in 94f76e7, we now get to a new crash later in the code:
```
root@qemu-cheri:~ # top
top: no termcap entry for a `vt100' terminal
CHERI cause: ExcCode: 0x01 RegNum: $c23 (length violation)
$c00: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c01: v:1 s:0 p:7fff807d b:0000000120275624 l:0000000000000020 o:0 t:0
$c02: v:1 s:0 p:7fff807d b:000000012033bc18 l:0000000000000008 o:0 t:0
$c03: v:1 s:0 p:7fff807d b:0000000120275624 l:0000000000000020 o:0 t:0
$c04: v:1 s:0 p:7fff807d b:000000012033bc28 l:0000000000000840 o:328 t:0
$c05: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:2023c0 t:0
$c06: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:41a360 t:0
$c07: v:1 s:0 p:7fff807d b:0000007fffffcf5c l:0000000000000001 o:0 t:0
$c08: v:1 s:0 p:7fff807d b:00000001202466b4 l:0000000000000004 o:0 t:0
$c09: v:1 s:0 p:7fff807d b:00000001202466bc l:0000000000000004 o:0 t:0
$c10: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:1202525e0 t:0
$c11: v:1 s:0 p:7fff807d b:0000007fff7ff000 l:00000000007ff3c0 o:0 t:0
$c12: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:1200ed5b0 t:0
$c13: v:1 s:0 p:00008055 b:0000007fffffd600 l:0000000000000060 o:0 t:0
$c14: v:1 s:0 p:7fff807d b:000000012026e500 l:0000000000000008 o:0 t:0
$c15: v:1 s:0 p:7fff807d b:000000012026e580 l:0000000000000008 o:0 t:0
$c16: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:12026e480 t:0
$c17: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:120010ca4 t:0
$c18: v:1 s:0 p:7fff807d b:0000000120258c20 l:0000000000000004 o:0 t:0
$c19: v:1 s:0 p:7fff807d b:0000000120246e64 l:0000000000000004 o:0 t:0
$c20: v:0 s:0 p:00000000 b:0000000000000000 l:0000000000000000 o:0 t:0
$c21: v:1 s:0 p:7fff807d b:0000007fffffd470 l:0000000000000010 o:0 t:0
$c22: v:1 s:0 p:7fff807d b:0000007fffffd458 l:0000000000000008 o:0 t:0
$c23: v:1 s:0 p:7fff807d b:0000007fffffe0e0 l:0000000000000100 o:e0 t:0
$c24: v:1 s:0 p:7fff807d b:0000007fffffd470 l:0000000000000010 o:0 t:0
$c26: v:1 s:0 p:7fff807d b:0000000000000000 l:0000010000000000 o:0 t:0
$c31: v:1 s:0 p:7ffe8017 b:0000000000000000 l:0000010000000000 o:120010dd4 t:0
Mar 22 13:24:44 qemu-cheri kernel: USER_CHERI_EXCEPTION: pid 556 tid 100056 (top), uid 0: CP2 fault (type 0x32)
Mar 22 13:24:45 qemu-cheri kernel: Trapframe Register Dump:
Mar 22 13:24:45 qemu-cheri kernel: zero: 0      at: 0x20        v0: 0x120275624v1: 0x1201e97c0
Mar 22 13:24:45 qemu-cheri kernel: a0: 0x12033bc18      a1: 0   a2: 0   a3: 0xffffffffffffffff
Mar 22 13:24:45 qemu-cheri kernel: a4: 0x1      a5: 0x30        a6: 0x30       a6: 0x100000000
Mar 22 13:24:45 qemu-cheri kernel: t0: 0x10     t1: 0   t2: 0   t3: 0x120270000
Mar 22 13:24:45 qemu-cheri kernel: t8: 0x10     t9: 0x12013e810 s0: 0x120260000s1: 0x12
Mar 22 13:24:45 qemu-cheri kernel: s2: 0x33     s3: 0x120260000 s4: 0   s5: 0x120260000
Mar 22 13:24:45 qemu-cheri kernel: s6: 0x4      s7: 0x20        k0: 0   k1: 0
Mar 22 13:24:45 qemu-cheri kernel: gp: 0x12025a6c0      sp: 0x7fe0a0    s8: 0x7fe0a0    ra: 0x120260000
Mar 22 13:24:45 qemu-cheri kernel: sr: 0x408084b3       mullo: 0x6e40   mulhi: 0badvaddr: 0x120010dd4
Mar 22 13:24:45 qemu-cheri kernel: cause: 0x48  pc: 0x120010dd4
Signal 34 (core dumped)
```

**Comment by rwatson** [comment]
Closing this issue; will open a fresh one on the new top problem.
